network/p2p/streams.go

```diff
@@ -93,51 +96,82 @@ func (n *streamManager) streamHandler(stream network.Stream) {
 	// no old stream
 	n.streams[stream.Conn().RemotePeer()] = stream
 	incoming := stream.Conn().Stat().Direction == network.DirInbound
-	n.handler(n.ctx, remotePeer, stream, incoming)
+	n.dispatch(n.ctx, remotePeer, stream, incoming)
 }
 
-// Connected is called when a connection is opened
-// for both incoming (listener -> addConn) and outgoing (dialer -> addConn) connections.
-func (n *streamManager) Connected(net network.Network, conn network.Conn) {
-	if conn.Stat().Direction == network.DirInbound && !n.allowIncomingGossip {
-		n.log.Debugf("ignoring incoming connection from %s", conn.RemotePeer().String())
-		return
+// dispatch the stream to the appropriate handler
+func (n *streamManager) dispatch(ctx context.Context, remotePeer peer.ID, stream network.Stream, incoming bool) {
+	if handler, ok := n.handlers[stream.Protocol()]; ok {
+		handler(ctx, remotePeer, stream, incoming)
+	} else {
+		n.log.Errorf("No handler for protocol %s, peer %s", stream.Protocol(), remotePeer)
+		_ = stream.Reset()
 	}
+}
 
-	remotePeer := conn.RemotePeer()
-	localPeer := n.host.ID()
+func (n *streamManager) peerWatcher(ctx context.Context, sub event.Subscription) {
+	defer sub.Close()
+	for e := range sub.Out() {
+		evt := e.(event.EvtPeerIdentificationCompleted)
+		conn := evt.Conn
 
-	// ensure that only one of the peers initiates the stream
-	if localPeer > remotePeer {
-		return
-	}
+		remotePeer := conn.RemotePeer()
+		localPeer := n.host.ID()
 
-	needUnlock := true
-	n.streamsLock.Lock()
-	defer func() {
-		if needUnlock {
+		if conn.Stat().Direction == network.DirInbound && !n.allowIncomingGossip {
+			n.log.Debugf("%s: ignoring incoming connection from %s", localPeer.String(), remotePeer.String())
+			continue
+		}
+
+		// ensure that only one of the peers initiates the stream
+		if localPeer > remotePeer {
+			n.log.Debugf("%s: ignoring a lesser peer ID %s", localPeer.String(), remotePeer.String())
```

Comment:
what's the idea with less peer ID ignoring here?

Reply:
both peers get notified. Without this they both will attempt to open duplex streams so we'll have duplicate data channels.